docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+
+### Element: Metric Propagation from Shoot to Seed
+To propagate metric data from individual shoot control planes to the seed level, the existing aggregate prometheus in
+the seed's garden namespace
+is leveraged. The aggregate prometheus already has access to shoot prometheus data, and implements a wildcard-based
+metric aggregation mechanism which, as long as certain naming convention is followed, aggregates the aforementioned new
+shoot prometheus metric without any need for changes.
+
+### Element: Prometheus Metrics Adapter
+A new component, named prometheus-metrics-adapter is added to seed clusters. It is an instance of the
+[kubernetes-sigs/prometheus-adapter]. It implements the
+Kubernetes Custom Metric API, exposing prometheus data in the form of K8s metrics selectively - controlled by
+configuration. A single replica is used, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
+
+### Element: Seed API Server Extension Service
+The prometheus-metrics-adapter is registered as an API extension to the seed API server. It is deployed with minimal
+deviations from the default configuration, to reduce future merge effort. As a result, it currently takes charge not only
+of the custom metric API, but also of the external metric API, preventing the deployment of alternative handlers for
+these API extension points. 
+
+### Element: HPA
+ShootKapi is scaled horizontally via HPA operating on the aforementioned average kube-apiserver request rate custom
+metric. Scale-in is delayed by a stabilization window. This avoids unnecessary flapping, the effects of which would
+be further amplified by VPA-driven evictions, and serves as a general precaution against HPA-VPA resonance,
+by ensuring that the two autoscalers operate with different frequency.
+
+The scaling threshold value for HPA is initially set to a conservative (lower-biased) estimate of the requests/replica
+ratio observed with the existing solution. It will be then gradually fine-tuned to the maximum value which does not
+impact quality of service.
+
+This proposal is the first stage of an incremental, two stage approach. It aims primarily at resolving the existing
+stability issues inherent to HVPA and enabling (but not executing) scaling efficiency optimizations. The second stage
+will be focused solely on further fine-tuning scaling efficiency to optimize ShootKapi compute cost over HVPA.
+Details follow in the [Discussion and Limitations](#discussion-and-limitations) section.
+
+### Element: VPA
+A standard VPA operates on ShootKapi. Its primary purpose is to improve resource efficiency over what's possible via
+HPA's coarse, replica-granularity scaling, by shrinking replicas down to match actual utilization.
+
+This improves over the existing solution by having vertical scaling active the entire time and not only during certain
+periods.
```

Comment:
Why significant? Hopefully soon it will become significant, after VPA starts applying updates in-place, without evicting pods. But right now, isn't it a relatively minor inconvenience (e.g. forcing Gardener to implement carry-over of resource values when reconciling Gardener components)?